plugins/woocommerce/src/Internal/StockNotifications/SettingsController.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace Automattic\WooCommerce\Internal\StockNotifications;
+
+/**
+ * Settings controller.
+ */
+class SettingsController extends \WC_Settings_Page {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+
+		// Add a 'Customer stock notifications' section to Products settings.
+		add_filter( 'woocommerce_get_sections_products', array( $this, 'add_customer_stock_notifications_section' ), 100, 1 );
+
+		// Add the Customer Stock Notifications settings.
+		add_filter( 'woocommerce_get_settings_products', array( $this, 'add_customer_stock_notifications_settings' ), 100, 2 );
+
+		// Display admin notices about incompatible settings combinations.
+		add_action( 'admin_notices', array( $this, 'output_admin_notices' ) );
+	}
+
+	/**
+	 * Add a 'Customer stock notifications' section to Products settings.
+	 *
+	 * @param array $sections Products settings sections.
+	 * @return array New Products settings sections.
+	 */
+	public function add_customer_stock_notifications_section( array $sections ): array {
```

Comment:
I'd avoid type hints when working with public hooks. These pesky errors often make it seem like Woo is crashing, but the truth is that it’s usually caused by external factors. Same applies to the method below.

Instead, we could do something like this and mitigate risks:
```php
if ( ! is_array( $sections ) ) {
  return $sections;
}
```

Reply:
Done via: https://github.com/woocommerce/woocommerce/pull/58202/commits/63b191e4d4b76e15ec5f97ee96b136e88b07d4d5